File: new-resource-detector/src/test/java/nl/haploid/octowight/service/EventHandlerServiceIT.java
package nl.haploid.octowight.service;

import nl.haploid.octowight.AbstractIT;
import nl.haploid.octowight.AtomChangeEvent;
import nl.haploid.octowight.JsonMapper;
import nl.haploid.octowight.TestData;
import nl.haploid.octowight.detector.MockResourceDetector;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.concurrent.ExecutionException;

import static org.junit.Assert.assertEquals;

public class EventHandlerServiceIT extends AbstractIT {

	@Autowired
	private EventHandlerService service;

	@Autowired
	private EventConsumerService consumerService;

	@Autowired
	private KafkaProducer<String, String> kafkaProducer;

	@Autowired
	private JsonMapper jsonMapper;

	@Test
	public void testHandleEventsNone() throws Exception {
		final long actualCount = service.handleEvents(10);
		assertEquals(0, actualCount);
	}

	@Test
	public void testHandleEvents() throws Exception {
		final String topic = TestData.topic();
		consumerService.setTopic(topic);
		final AtomChangeEvent event1 = TestData.atomChangeEvent(MockResourceDetector.ATOM_TYPE);
		final AtomChangeEvent event2 = TestData.atomChangeEvent(MockResourceDetector.ATOM_TYPE);
		final AtomChangeEvent event3 = TestData.atomChangeEvent("jack");
		sendMessage(topic, event1);
		sendMessage(topic, event2);
		sendMessage(topic, event3);
		final long actualCount = service.handleEvents(10);
		assertEquals(2, actualCount);
	}

	public void sendMessage(final String topic, final AtomChangeEvent event) throws InterruptedException, ExecutionException {
		final String message = jsonMapper.serialize(event);
		log.debug(String.format("Send message: %s", message));
		final ProducerRecord<String, String> record = new ProducerRecord<>(topic, message);
		kafkaProducer.send(record).get();
	}
}

File: new-resource-detector/src/test/scala/nl/haploid/octowight/service/EventHandlerServiceIT.scala
package nl.haploid.octowight.service

import java.util.concurrent.TimeUnit

import nl.haploid.octowight.detector.MockResourceDetector
import nl.haploid.octowight.kafka.KafkaProducerFactory
import nl.haploid.octowight.{AbstractIT, AtomChangeEvent, JsonMapper, TestData}
import org.apache.kafka.clients.producer.ProducerRecord
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.util.ReflectionTestUtils

class EventHandlerServiceIT extends AbstractIT {
  @Autowired private[this] val eventHandlerService: EventHandlerService = null
  @Autowired private[this] val eventConsumerService: EventConsumerService = null
  @Autowired private[this] val kafkaProducerFactory: KafkaProducerFactory = null
  @Autowired private[this] val jsonMapper: JsonMapper = null

  private[this] lazy val kafkaProducer = kafkaProducerFactory.kafkaProducer

  private[this] val topic = TestData.topic

  override def beforeEach() = {
    super.beforeEach()
    ReflectionTestUtils.setField(eventConsumerService, "topic", topic)
  }

  "Event handler service" should "handle no events" in {
    val actualCount = eventHandlerService.detectNewResources(10)
    actualCount should be(0)
  }

  "Event handler service" should "handle some events" in {
    val event1 = TestData.atomChangeEvent(MockResourceDetector.AtomCategory)
    val event2 = TestData.atomChangeEvent(MockResourceDetector.AtomCategory)
    val event3 = TestData.atomChangeEvent("jack")
    sendMessage(topic, event1)
    sendMessage(topic, event2)
    sendMessage(topic, event3)
    val actualCount = eventHandlerService.detectNewResources(10)
    actualCount should be(2)
  }

  def sendMessage(topic: String, event: AtomChangeEvent) = {
    val message = jsonMapper.serialize(event)
    log.debug(s"Send message to $topic: $message")
    val record = new ProducerRecord[String, String](topic, message)
    kafkaProducer.send(record).get(5, TimeUnit.SECONDS)
  }
}
